lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -887,9 +887,14 @@ static ExprSolution solveExpr(Expr *expr, SmallPtrSetImpl<Expr *> &seenVars,
         .Case<VarExpr>([&](auto *expr) {
           if (solvedExprs.contains(expr->constraint)) {
             auto solution = solvedExprs[expr->constraint];
-            if (expr->upperBound)
-              expr->upperBoundSolution = solvedExprs[expr->upperBound].second;
-
+            // If we've solved the upper bound already, store the solution.
+            // This will be explicitly solved for later if not computed as
+            // part of the solving that resolved this constraint.
+            // This should only happen if somehow the constraint is
+            // solved before visiting this expression, so that our upperBound
+            // was not added to the worklist such that it was handled first.
+            if (expr->upperBound && solvedExprs.contains(expr->upperBound))
```

Comment:
I'm not sure we don't need to handle this re:seenVars and checkCycles/etc.?
Or if constraint expressions can be re-used such that we could have solution for the constraint w/o having put the upper bound on the worklist first?